xen/arch/arm/coproc/coproc.c

```diff
@@ -242,218 +371,240 @@ static int coproc_detach_from_domain(struct domain *d,
 
     coproc_deinit_vcoproc(vcoproc);
 
-    printk("Destroyed vcoproc \"%s\" for dom%u\n",
-            dev_path(coproc->dev), d->domain_id);
+    dprintk(XENLOG_INFO, "Destroyed vcoproc \"%s\" for dom%u\n",
+            dev_path(mcoproc->dev), d->domain_id);
 
 out:
     spin_unlock(&coprocs_lock);
 
     return ret;
 }
 
-bool_t coproc_is_attached_to_domain(struct domain *d, const char *path)
-{
-    struct coproc_device *coproc;
-    bool_t is_created;
-
-    coproc = coproc_find_by_path(path);
-    if ( !coproc )
-        return false;
-
-    spin_lock(&coprocs_lock);
-    is_created = coproc_is_created_vcoproc(d, coproc);
-    spin_unlock(&coprocs_lock);
-
-    return is_created;
-}
-
-struct coproc_device *coproc_alloc(struct dt_device_node *np,
-                                   const struct coproc_ops *ops)
+static int mcoproc_acquire_mmios(struct mcoproc_device *mcoproc,
+                                 const struct pcoproc_desc *desc)
 {
-    struct coproc_device *coproc;
-    struct device *dev = &np->dev;
-    unsigned int num_irqs, num_mmios;
-    int i, ret;
-
-    coproc = xzalloc(struct coproc_device);
-    if ( !coproc )
-    {
-        printk("Failed to allocate coproc_device for \"%s\"\n",
-               dev_path(dev));
-        return ERR_PTR(-ENOMEM);
-    }
-    coproc->dev = dev;
+    int i;
+    int ret = 0;
+    u32 num_mmios;
+    struct dt_device_node *np = dev_to_dt(mcoproc->dev);
 
     num_mmios = dt_number_of_address(np);
 
     if ( !num_mmios )
     {
         printk("Failed to find at least one mmio for \"%s\"\n",
-               dev_path(coproc->dev));
+               dev_path(mcoproc->dev));
+        ret = -ENODEV;
+        goto out;
+    }
+
+    if ( num_mmios != desc->p_mmio_num )
+    {
+        printk("MMIO num mistmatch for \"%s\"\n",
+               dev_path(mcoproc->dev));
         ret = -ENODEV;
         goto out;
     }
 
-    coproc->mmios = xzalloc_array(struct mmio, num_mmios);
-    if ( !coproc->mmios )
+    mcoproc->mmios = xzalloc_array(struct mcoproc_mmio, num_mmios);
+    if ( !mcoproc->mmios )
     {
         printk("Failed to allocate %d mmio(s) for \"%s\"\n", num_mmios,
-               dev_path(coproc->dev));
+               dev_path(mcoproc->dev));
         ret = -ENOMEM;
         goto out;
     }
 
-    for ( i = 0; i < num_mmios; ++i )
+    for ( i =0; i < num_mmios; i++)
     {
-        struct mmio *mmio = &coproc->mmios[i];
-        ret = dt_device_get_address(np, i, &mmio->addr, &mmio->size);
+        struct mcoproc_mmio *m_mmio = &mcoproc->mmios[i];
+        struct pcoproc_mmio *p_mmio = &desc->p_mmio[i];
+        int index = dt_property_match_string(np, "reg-names", p_mmio->name);
+
+        if (index < 0)
+        {
+            if ( num_mmios > 1 )
+            {
+                printk("More than one mmio and failed name match for %s\n",
```

Comment:
It obviously says that coproc has more than one mmio, and mmios name matching failed.